file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingPath.kt

```diff
@@ -0,0 +1,276 @@
+package com.pkware.file.fileAttributeCaching
+
+import com.pkware.file.forwarding.ForwardingPath
+import java.nio.file.FileSystem
+import java.nio.file.LinkOption
+import java.nio.file.Path
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.attribute.DosFileAttributes
+import java.nio.file.attribute.PosixFileAttributes
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+// Set the cache preservation duration to 5 seconds as most file operations complete in this time.
+internal const val CACHE_PRESERVATION_TIME_SECONDS: Long = 5
+
+/**
+ * A [Path] instance that supports caching of [BasicFileAttributes] and other classes that extend it such as
+ * [DosFileAttributes] and [PosixFileAttributes].
+ *
+ * @param fileSystem the [FileSystem] associated with this [FileAttributeCachingPath] instance.
+ * @param delegate the [Path] to forward calls to if needed.
+ * @param cacheTimeout the amount of time (in seconds) allowed before flushing this [FileAttributeCachingPath]'s cache.
+ * The default is [CACHE_PRESERVATION_TIME_SECONDS].
+ */
+internal class FileAttributeCachingPath(
+    private val fileSystem: FileSystem,
+    internal val delegate: Path,
+    internal val cacheTimeout: Long = CACHE_PRESERVATION_TIME_SECONDS
+) : ForwardingPath(delegate) {
+
+    /**
+     * Instance of [ExpirableCache] that maps attribute class strings (*, dos:*, posix:*) to [BasicFileAttributes].
+     */
+    private val attributeCache = ExpirableCache<String, BasicFileAttributes?>(
+        TimeUnit.SECONDS.toMillis(cacheTimeout)
+    )
+
+    override fun getFileSystem(): FileSystem = fileSystem
+
+    override fun resolve(other: Path): Path {
+        // Make sure if other is a FileAttributeCachingPath that we pass along other's
+        // delegate rather than other itself.
+        val resolvedDelegatePath = if (other is FileAttributeCachingPath) {
+            delegate.resolve(other.delegate)
+        } else {
+            delegate.resolve(other)
+        }
+
+        // TODO should we copy the cache here?
+        return FileAttributeCachingPath(fileSystem, resolvedDelegatePath, cacheTimeout)
```

Comment:
I think the answer to your TODO is yes, we should copy the cache.

However, you might want to double check these calls (and some of these other methods) for infinite recursion.

For example:

```
val resolvedDelegatePath = if (other is FileAttributeCachingPath) {
            delegate.resolve(other.delegate)
}
```

Reply:
not going to answer that right now and leave it for the meeting because what if we arent dealing with a path that we dont care about caching its attributes or the path is not even remotely related to the file we are scanning or the path is not a file?

Reply:
im only copying the cache for path operations that operate only with the path itself and asking about the rest- it does not make sense to copy the cache for every operation

Reply:
also for that call in question, im just passing the other.delegate to `this` path's delegate resolve- there should be lower level logic thats handled by the wrapped filesystem/path that prevents infinite recursion

Reply:
What I think I failed to articulate is that we probably want to deny the opportunity for `FileAttributeCachingPath` to ever be constructed with a `FileAttributeCachingPath` delegate.

Reply:
then what should happen in this case? i just throw an exception? not even sure how to test that

Reply:
okay so we will wrap everything in a cache path but only copy attributes on operations that affect the actual path itself (not any other path) got it